src/Workspaces/Core/Portable/Workspace/Solution/SourceGeneratedDocumentState.cs

```diff
@@ -0,0 +1,110 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Threading;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis
+{
+    internal class SourceGeneratedDocumentState : DocumentState
+    {
+        public static SourceGeneratedDocumentState Create(
+            GeneratedSourceResult generatedSourceResult,
+            DocumentId documentId,
+            ISourceGenerator sourceGenerator,
+            HostLanguageServices languageServices,
+            SolutionServices solutionServices)
+        {
+            var textAndVersion = TextAndVersion.Create(generatedSourceResult.SourceText, VersionStamp.Create());
+            ValueSource<TextAndVersion> textSource = new ConstantValueSource<TextAndVersion>(textAndVersion);
+
+            var tree = generatedSourceResult.SyntaxTree;
+
+            // Since the tree is coming directly from the generator, this tree is strongly held so GetRoot() doesn't need a CancellationToken.
+            var root = tree.GetRoot(CancellationToken.None);
+            Contract.ThrowIfNull(languageServices.SyntaxTreeFactory, "We should not have a generated syntax tree for a language that doesn't support trees.");
+
+            if (languageServices.SyntaxTreeFactory.CanCreateRecoverableTree(root))
+            {
+                // We will only create recoverable text if we can create a recoverable tree, otherwise we'd have a
+                // tree wrapping the original text and that may cause identity issues.
+                textSource = CreateRecoverableText(textAndVersion, solutionServices);
+                tree = languageServices.SyntaxTreeFactory.CreateRecoverableTree(
+                    documentId.ProjectId,
+                    filePath: null,
```

Comment:
slightly concerning.  are you ok with ths?  should we doc why this is appropriate?

Reply:
Oops good catch.

Reply:
(and fixing tests that didn't catch this...)